velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+spill io cost but also speed up the restore procedure. For un-spilled
+partitions, we can process them all together with the normal fast execution
+path. For spilled partitions, each will have its own set of files on disk and
+will be restored separately. The columns used to calculate the partition number
+are specified by the operator based on the spilling algorithm. Take hash
+aggregation for example, it sets the partition columns to the group by keys.
+This ensures the rows with the same group by key will spill to the same
+partition and then be restored together.
+
+**spill partition selection**: we select partitions with the most spillable data to
+spill initially, and then stick with those partitions if they have sufficient
+memory to free up. The spillable data is measured as the memory bytes occupied
+by the rows in the row container. For operators which need sorted spilling, we
+might need to avoid spilling from partitions which have a very small amount of
+spillable data even though it has already been spilled. The reason is that the
+sorted spilling requires that each spill file contains a single sorted run
+which means we can’t continuously append data to the same spill file. There is
+no such restriction for non-sorted spilling such as for the hash build
+operator.
+
+**spill data sort**: the operator that needs sorted spilling will run sort merge
+on top of the spill files plus the in-memory data to produce the final result.
+The in-memory data will also be sorted as one of the sorted runs. The Spiller
+sorts the rows by partition columns with a set of comparison flags specified by
+the operator. For example, the order by operator needs to ensure the sorted
+runs follow the same sort order as specified by the query plan node.
+
+**spill data io**: the Spiller handles the interaction with the storage system
+through `SpillFileList <https://github.com/facebookincubator/velox/blob/main/velox/exec/Spill.h#L155>`_ and `SpillFile <https://github.com/facebookincubator/velox/blob/main/velox/exec/Spill.h#L58>`_ objects as discussed below. It manages the
+lifecycle of a spill file from creation, write, read and deletion. The spill
+writes are offloaded to a dedicated IO executor and each spill partition write
+is a thread execution unit. The spill reads are executed in the driver
+executor. Both read and write are synchronous IO operations.
+
+The Spiller provides the following spilling APIs for operators to use:
+
+**spill with targets**: the operator specifies the number of rows and bytes to
+spill as the target. The Spiller will select a number of partitions to spill to
+meet the spilling target. The spilling is running internally and the function
+will return to the user after the spilling completes. The spilling process is
+not controlled by the operator but it can check which partitions have been
+spilled and how much data has been spilled through spill stats API.
```

Comment:
> through spill stats API.

The reader would expect the stats API next, but the document shows a 'spill' API. Would it make sense to add the stats API here?